pubsub/consumer.go

```diff
@@ -57,27 +65,48 @@ func NewConsumer[Request any, Response any](client redis.UniversalClient, stream
 	if streamName == "" {
 		return nil, fmt.Errorf("redis stream name cannot be empty")
 	}
-	consumer := &Consumer[Request, Response]{
+	return &Consumer[Request, Response]{
 		id:          uuid.NewString(),
 		client:      client,
 		redisStream: streamName,
 		redisGroup:  streamName, // There is 1-1 mapping of redis stream and consumer group.
 		cfg:         cfg,
-	}
-	return consumer, nil
+		terminating: atomic.Bool{},
+		signals:     make(chan os.Signal, 1),
+	}, nil
 }
 
 // Start starts the consumer to iteratively perform heartbeat in configured intervals.
 func (c *Consumer[Request, Response]) Start(ctx context.Context) {
 	c.StopWaiter.Start(ctx, c)
+	c.listenForInterrupt()
 	c.StopWaiter.CallIteratively(
 		func(ctx context.Context) time.Duration {
+			if !c.terminating.Load() {
+				log.Trace("Consumer is terminating, stopping heartbeat update")
+				return time.Hour
+			}
 			c.heartBeat(ctx)
 			return c.cfg.KeepAliveTimeout / 10
 		},
 	)
 }
 
+// listenForInterrupt launches a thread that notifies the channel when interrupt
+// is received.
+func (c *Consumer[Request, Response]) listenForInterrupt() {
```

Comment:
there is already signal.Notify in both nitro and nitro-val.
This will take care of catching the sigint. Nitro will issue StopAndWait and I think nitro-val won't and it needs to be added (please verify)
Then all you need to do is delete heartbeat in Consumer's StopAndWait. By doing it after calling parent's StopAndWait you can also tell there will be no race condition with the heartbeat thread, because that thread has already stopped. (likewise no need for "terminating" as that's already handled by the stopwaiter

Reply:
Another note: if you need a usable context after calling stopwaiter.StopAndWait, you can use:
GetParentContext